File: 0078-subsets/0078-subsets.java
class Solution {
    List<List<Integer>> ans = new ArrayList<>();

    public void dfs(int[] nums, List<Integer> temp, int start){
         ans.add(new ArrayList<>(temp)); // Push the current subset into the resultant array

        for(int i = start; i < nums.length; i++){
            temp.add(nums[i]); // Add the current element to consider the subsets corresponding to it
            dfs(nums, temp, i+1); // Generate subsets for this array
            temp.remove(temp.size()-1);
        }
    }
    
    
    public List<List<Integer>> subsets(int[] nums) {
        
        List<Integer> temp = new ArrayList<>();

        dfs(nums, temp, 0);

        return ans;
    }
}

File: 0078-subsets/0078-subsets.cpp
class Solution {
public:
    vector<vector<int>> ans;
    void dfs(vector<int>& nums, int idx, vector<int> temp){
        ans.push_back(temp);

        for(int i = idx; i < nums.size(); i++){
            temp.push_back(nums[i]);
            dfs(nums, i+1, temp);
            temp.pop_back();
        }
    }
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<int> temp;
        dfs(nums,0,temp);

        return ans;
    }
};